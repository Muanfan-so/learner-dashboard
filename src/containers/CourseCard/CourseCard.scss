.course-card {
  .card {
    overflow: hidden;

    .pgn__card-image-cap {
      border-bottom-left-radius: 0 !important;
    }
  }

  .course-card-banners {
    > .alert {
      border-top-left-radius: 0;
      border-top-right-radius: 0;
    }
  }
}