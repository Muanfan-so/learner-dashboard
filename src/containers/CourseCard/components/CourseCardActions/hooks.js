import { handleEvent } from 'data/services/segment/utils';
import { eventNames } from 'data/services/segment/constants';

export const useCourseCardActionData = () => {
  const trackUpgradeClick = () => {
    handleEvent(eventNames.upgradeCourse, {
      pageName: 'learner_home',
      linkType: 'button',
      linkCategory: 'green_upgrade',
    });
  };

  return {
    trackUpgradeClick,
  };
};

export default useCourseCardActionData;
