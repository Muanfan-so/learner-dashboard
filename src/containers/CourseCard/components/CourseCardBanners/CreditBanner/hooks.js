import { StrictDict } from 'utils';
import { hooks as appHooks } from 'data/redux';

import ApprovedContent from './views/ApprovedContent';
import EligibleContent from './views/EligibleContent';
import MustRequestContent from './views/MustRequestContent';
import PendingContent from './views/PendingContent';
import RejectedContent from './views/RejectedContent';

export const statusComponents = StrictDict({
  pending: PendingContent,
  approved: ApprovedContent,
  rejected: RejectedContent,
});

export const useCreditBannerData = (cardId) => {
  const credit = appHooks.useCardCreditData(cardId);
  const { supportEmail } = appHooks.usePlatformSettingsData();
  if (!credit.isEligible) { return null; }

  const { error, purchased, requestStatus } = credit;
  let ContentComponent = EligibleContent;
  if (purchased) {
    if (requestStatus == null) {
      ContentComponent = MustRequestContent;
    } else if (Object.keys(statusComponents).includes(requestStatus)) {
      ContentComponent = statusComponents[requestStatus];
    }
    // Current behavior is to show Elligible State if unknown request status is returned
  }
  return {
    ContentComponent,
    error,
    supportEmail,
  };
};

export default {
  useCreditBannerData,
};
