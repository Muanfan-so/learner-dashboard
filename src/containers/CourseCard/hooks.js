import { useIntl } from '@edx/frontend-platform/i18n';
import { hooks as appHooks } from 'data/redux';

export const useCardData = ({ cardId }) => {
  const { formatMessage } = useIntl();
  const { title, bannerImgSrc } = appHooks.useCardCourseData(cardId);
  const { isEnrolled } = appHooks.useCardEnrollmentData(cardId);

  return {
    isEnrolled,
    title,
    bannerImgSrc,
    formatMessage,
  };
};

export default useCardData;
