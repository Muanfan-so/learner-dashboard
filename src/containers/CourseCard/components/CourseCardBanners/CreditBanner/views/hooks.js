import React from 'react';

import { StrictDict } from 'utils';
import { apiHooks } from 'hooks';

import * as module from './hooks';

export const state = StrictDict({
  creditRequestData: (val) => React.useState(val), // eslint-disable-line
});

export const useCreditRequestData = (cardId) => {
  const [requestData, setRequestData] = module.state.creditRequestData(null);
  const createCreditApiRequest = apiHooks.useCreateCreditRequest(cardId);
  const createCreditRequest = (e) => {
    e.preventDefault();
    createCreditApiRequest().then(setRequestData);
  };
  return { requestData, createCreditRequest };
};

export default {
  useCreditRequestData,
};
