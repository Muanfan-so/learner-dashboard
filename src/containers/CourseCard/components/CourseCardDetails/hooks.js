import { useIntl } from '@edx/frontend-platform/i18n';
import { hooks as appHooks } from 'data/redux';

import * as hooks from './hooks';
import messages from './messages';

export const useAccessMessage = ({ cardId }) => {
  const { formatMessage, formatDate } = useIntl();
  const enrollment = appHooks.useCardEnrollmentData(cardId);
  const courseRun = appHooks.useCardCourseRunData(cardId);
  if (!courseRun.isStarted) {
    if (!courseRun.startDate) { return null; }
    const startDate = formatDate(courseRun.startDate);
    return formatMessage(messages.courseStarts, { startDate });
  }
  if (enrollment.isEnrolled) {
    const { isArchived, endDate } = courseRun;
    const {
      accessExpirationDate,
      isAuditAccessExpired,
    } = enrollment;

    if (enrollment.isAudit && accessExpirationDate) {
      return formatMessage(
        isAuditAccessExpired ? messages.accessExpired : messages.accessExpires,
        { accessExpirationDate: formatDate(accessExpirationDate) },
      );
    }
    if (!endDate) { return null; }
    return formatMessage(
      isArchived ? messages.courseEnded : messages.courseEnds,
      { endDate: formatDate(endDate) },
    );
  }
  return null;
};

export const useCardDetailsData = ({ dispatch, cardId }) => {
  const { formatMessage } = useIntl();
  const providerName = appHooks.useCardProviderData(cardId).name;
  const { courseNumber } = appHooks.useCardCourseData(cardId);
  const {
    isEntitlement,
    isFulfilled,
    canChange,
  } = appHooks.useCardEntitlementData(cardId);

  const openSessionModal = appHooks.useUpdateSelectSessionModalCallback(dispatch, cardId);

  return {
    providerName: providerName || formatMessage(messages.unknownProviderName),
    accessMessage: hooks.useAccessMessage({ cardId }),
    isEntitlement,
    isFulfilled,
    canChange,
    openSessionModal,
    courseNumber,
    changeOrLeaveSessionMessage: formatMessage(messages.changeOrLeaveSessionButton),
  };
};

export default useCardDetailsData;
