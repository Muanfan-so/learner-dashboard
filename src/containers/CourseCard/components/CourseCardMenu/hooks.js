import React from 'react';
import { StrictDict } from 'utils';

import track from 'tracking';
import { reduxHooks } from 'hooks';

import * as module from './hooks';

export const state = StrictDict({
  isUnenrollConfirmVisible: (val) => React.useState(val), // eslint-disable-line
  isEmailSettingsVisible: (val) => React.useState(val), // eslint-disable-line
});

export const useUnenrollData = () => {
  const [isVisible, setIsVisible] = module.state.isUnenrollConfirmVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const useEmailSettings = () => {
  const [isVisible, setIsVisible] = module.state.isEmailSettingsVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const useHandleToggleDropdown = (cardId) => {
  const eventName = track.course.courseOptionsDropdownClicked;
  const trackCourseEvent = reduxHooks.useTrackCourseEvent(eventName, cardId);
  return (isOpen) => {
    if (isOpen) { trackCourseEvent(); }
  };
};

export const useCourseCardMenu = (cardId) => {
  const { courseName } = reduxHooks.useCardCourseData(cardId);
  const { isEnrolled, isEmailEnabled } = reduxHooks.useCardEnrollmentData(cardId);
  const { twitter, facebook } = reduxHooks.useCardSocialSettingsData(cardId);
  const { isMasquerading } = reduxHooks.useMasqueradeData();
  const { isEarned } = reduxHooks.useCardCertificateData(cardId);
  const handleTwitterShare = reduxHooks.useTrackCourseEvent(
    track.socialShare,
    cardId,
    'twitter',
  );
  const handleFacebookShare = reduxHooks.useTrackCourseEvent(
    track.socialShare,
    cardId,
    'facebook',
  );

  const showUnenrollItem = isEnrolled && !isEarned;
  const showDropdown = showUnenrollItem || isEmailEnabled || facebook.isEnabled || twitter.isEnabled;

  return {
    courseName,
    isMasquerading,
    isEmailEnabled,
    showUnenrollItem,
    showDropdown,
    facebook,
    twitter,
    handleTwitterShare,
    handleFacebookShare,
  };
};
