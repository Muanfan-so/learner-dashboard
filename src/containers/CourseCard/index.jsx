import React from 'react';
import PropTypes from 'prop-types';

// import PropTypes from 'prop-types';
import { Card } from '@edx/paragon';

import useCardData from './hooks';

import RelatedProgramsBadge from './components/RelatedProgramsBadge';
import CourseCardMenu from './components/CourseCardMenu';
import {
  CourseBanner,
  CertificateBanner,
  EntitlementBanner,
} from './components/Banners';
import CourseCardActions from './components/CourseCardActions';
import messages from './messages';
import CourseCardDetails from './components/CourseCardDetails';

import './CourseCard.scss';

export const CourseCard = ({ cardId }) => {
  const {
    isEnrolled, title, bannerUrl, formatMessage,
  } = useCardData({
    cardId,
  });
  return (
    <div className="mb-4.5 course-card" data-testid="CourseCard">
      <Card orientation="horizontal">
        <div className="d-flex flex-column w-100">
          <div className="d-flex">
            <Card.ImageCap
              src={bannerUrl}
              srcAlt={formatMessage(messages.bannerAlt)}
            />
            <Card.Body>
              <Card.Header
                title={<span data-testid="CourseCardTitle">{title}</span>}
                actions={<CourseCardMenu cardId={cardId} />}
              />
              <Card.Section>
                <CourseCardDetails cardId={cardId} />
              </Card.Section>
              <Card.Footer
                orientation="vertical"
                textElement={<RelatedProgramsBadge cardId={cardId} />}
              >
                <CourseCardActions cardId={cardId} />
              </Card.Footer>
            </Card.Body>
          </div>

          <div className="course-card-banners" data-testid="CourseCardBanners">
            <CourseBanner cardId={cardId} />
            <EntitlementBanner cardId={cardId} />
            {isEnrolled && <CertificateBanner cardId={cardId} />}
          </div>
        </div>
      </Card>
    </div>
  );
};

CourseCard.propTypes = {
  cardId: PropTypes.string.isRequired,
};

CourseCard.defaultProps = {};

export default CourseCard;
