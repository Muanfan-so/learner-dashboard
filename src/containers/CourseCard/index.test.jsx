import React from 'react';
import { shallow } from 'enzyme';

import CourseCard from '.';
import hooks from './hooks';

jest.mock('./hooks', () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock('./components/RelatedProgramsBadge', () => 'RelatedProgramsBadge');
jest.mock('./components/CourseCardMenu', () => 'CourseCardMenu');
jest.mock('./components/Banners', () => ({
  CourseBanner: () => 'CourseBanner',
  CertificateBanner: () => 'CertificateBanner',
  EntitlementBanner: () => 'EntitlementBanner',
}));
jest.mock('./components/CourseCardActions', () => 'CourseCardActions');
jest.mock('./components/CourseCardDetails', () => 'CourseCardDetails');

const dataProps = {
  title: 'hooks.title',
  bannerUrl: 'hooks.bannerUrl',
  formatMessage: jest.fn(msg => ({ formatted: msg })),
};

const courseNumber = 'test-course-number';

describe('CourseCard component', () => {
  test('snapshot', () => {
    hooks.mockReturnValueOnce(dataProps);
    expect(shallow(<CourseCard courseNumber={courseNumber} />)).toMatchSnapshot();
    expect(hooks).toHaveBeenCalledWith({ courseNumber });
  });
});
