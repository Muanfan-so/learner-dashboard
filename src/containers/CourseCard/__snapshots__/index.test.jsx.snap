// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseCard component snapshot 1`] = `
<div
  className="mb-4.5 course-card"
  data-testid="CourseCard"
>
  <Card
    orientation="horizontal"
  >
    <div
      className="d-flex flex-column w-100"
    >
      <div
        className="d-flex"
      >
        <Card.ImageCap
          src="hooks.bannerUrl"
          srcAlt={
            Object {
              "formatted": Object {
                "defaultMessage": "Course thumbnail",
                "description": "Course card banner alt-text",
                "id": "learner-dash.courseCard.bannerAlt",
              },
            }
          }
        />
        <Card.Body>
          <Card.Header
            actions={
              <CourseCardMenu
                cardId="test-card-id"
              />
            }
            title={
              <span
                data-testid="CourseCardTitle"
              >
                hooks.title
              </span>
            }
          />
          <Card.Section>
            <CourseCardDetails
              cardId="test-card-id"
            />
          </Card.Section>
          <Card.Footer
            orientation="vertical"
            textElement={
              <RelatedProgramsBadge
                cardId="test-card-id"
              />
            }
          >
            <CourseCardActions
              cardId="test-card-id"
            />
          </Card.Footer>
        </Card.Body>
      </div>
      <div
        className="course-card-banners"
        data-testid="CourseCardBanners"
      >
        <CourseBanner
          cardId="test-card-id"
        />
        <EntitlementBanner
          cardId="test-card-id"
        />
        <CertificateBanner
          cardId="test-card-id"
        />
      </div>
    </div>
  </Card>
</div>
`;
